onnxruntime/core/graph/function_utils.cc

```diff
@@ -26,34 +26,52 @@ static int GetVersionForDomain(const std::string& domain, const M<std::string, i
   return it->second;
 }
 
-std::unique_ptr<ONNX_NAMESPACE::OpSchema>
-CreateSchema(const Graph& graph,
-             const IndexedSubGraph& nodes_to_fuse) {
+std::unique_ptr<ONNX_NAMESPACE::OpSchema> CreateSchema(
+    const Graph& graph,
+    const IndexedSubGraph& nodes_to_fuse, bool allow_anytype_tensor) {
   const auto* meta_def = nodes_to_fuse.GetMetaDef();
-  auto op_schema = std::make_unique<ONNX_NAMESPACE::OpSchema>();
-  op_schema->SetName(meta_def->name);
+
+  using ONNX_NAMESPACE::OpSchema;
+  auto op_schema = std::make_unique<OpSchema>(meta_def->name, __FILE__, __LINE__);
   op_schema->SetDomain(meta_def->domain);
   op_schema->SetDoc(meta_def->doc_string);
   op_schema->SinceVersion(meta_def->since_version);
 
   if (meta_def->type_and_shape_inference_function) {
     op_schema->TypeAndShapeInferenceFunction(meta_def->type_and_shape_inference_function);
   }
+  if (allow_anytype_tensor) {
+    // we manually check if the node is supported in the EP, so if we ever create a node in our custom domain
+    // we know it's supported. due to that the type constraints here don't really matter - they just need to be cover
+    // all valid values. extra types for a specific input or output don't hurt.
+    op_schema->TypeConstraint("T_ANY", ONNX_NAMESPACE::OpSchema::all_tensor_types_with_bfloat(),
+                              "any type");
+    // match `input` and `output`, It's used when a Op dev defined a `T` constraint
+    op_schema->TypeConstraint("T", ONNX_NAMESPACE::OpSchema::all_tensor_types_with_bfloat(),
+                              "any type");
+  }
```

Comment:
Currently,  we register kernels by static and ususally have a type-constraint for input/output with a named-constraint "T",

ie, 
```
ONNX_OPERATOR_KERNEL_EX(QLinearSoftmax, kDynamicDomainByCreate, 1, kXnnpackExecutionProvider,
                        KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<uint8_t>()),
                        Softmax);
```
Without  "T" constraint, w can't have constraint in kernel-def either.

Reply:
Can we use 'TAny' in the kernel registration instead? The kernel def needs to match the schema in terms of constraint name, so I think it's reasonable for the schema creation to define the name to use for this 'any' constraint (vs. doing it the other way around and attempting to match the kernel def constraint name in the schema creation). 

The string to use for the 'any' constraint should be defined somewhere as a constant so an EP using this facility can use that instead of hardcoding 'TAny' in various places. FWIW 'TAny' or 'Tany' would be slightly more consistent with the ONNX schemas which have things like 'T1' and 'Tind' (and not 'T_1' or 'T_ind').

Note that still should allow the kernel def to use a subset of types from the 'TAny' constraint if we really need to do that (TBD how well that would work with a minimal build - may be better to handle types inside the kernel implementation). 

e.g. https://github.com/onnx/onnx/blob/main/docs/Operators.md#Resize specifies many different types are allowed for the 'T' constraint. The kernel registrations for the CPU EP have separate kernels for each type. https://github.com/microsoft/onnxruntime/blob/b83ea3c2ff305069db779c4b5f980cfd38915207/onnxruntime/core/providers/cpu/tensor/resize.cc#L8-L30

Reply:
Maybe we can't use TAny in Kernel def constraint.
We use TAny in schema's type-constraint is to relax the type-check when matching kernels. We only need one schema to  match for many typed kernels.

And we want to use 'T' to represent the input type of a Op-kernel. Even a kernel has multiple inputs sometime, like QConv has 9 inputs with type of (uint8/int8, float, int32), but we usually only care the first input, so we applied 'T' on the first input-type-constraint.

If we use TAny in the kernel-def, we have to use something 
```
KernelDefBuilder().TypeConstraint("T", {DataTypeImpl::GetTensorType<float>(),
                                                                DataTypeImpl::GetTensorType<int32>() ,
                                                                DataTypeImpl::GetTensorType<uint8>() ,
                                                                DataTypeImpl::GetTensorType<int8>() ,
 }), 
```
we have to put all types-allowed among all input-tensor

Please correct me if I am wrong.

Reply:
Having both 'T' and 'TAny' is somewhat opaque and potentially confusing for a user to understand. They need to do type aggregation in one place for the 'TAny' constraint but it's optional in the other for the 'T' constraint. The relationship between 'T' and 'TAny' isn't that clear from the naming either. 

If the EP does manual type validation for the kernel I don't think the kernel registration requires any constraints does it? Looking at the CPU Resize kernel registrations they only specify the T1 constraint and not the T2.

It's also relatively simple for the OpKernel::Compute method to check the input data type and call a type specific implementation, so the kernel registration having a type constraint for the input data seems optional to me. It may also be possible to specific a single type for an input/output parameter with the inline type constraint and have it match against the 'TAny' constraint - would need to check.

e.g. if the kernel input was called 'X', we create the schema with 'TAny' as the named constraint, but declare the kernel with an inline type constraint for 'X' (i.e. `.TypeConstraint("X", DataTypeImpl::GetTensorType<float>())`)

I think that gives at least a couple of options. We should also keep in mind that the EP can/should do manual type constraint checking prior to creating the node that uses the generated schema, and the EP can potentially have different kernels for different types (e.g. it could have an xnnpack.QLinearSoftmax_uint8 kernel and an xnnpack.QLinearSoftmax_int8 kernel) if they _really really_ needed that.

- Aggregate all types with a single type constraint. 'TAny' doesn't seem like a clear enough name for this usage though and maybe something like 'TAggregatedTypes' would help. However if specifying a TypeConstraint in the kernel registration is not required, the kernel registration wouldn't need to use the name at all.
  -  and if we can do the inline constraint to limit a particular input/output to a single type that may provide enough functionality without over complicating usage
- Have a separate constraint for every input/output so that there's no aggregation across them, with names like 'TInput0', 'TInput1', 'TOutput0' etc.  That has clearer usage but costs more in the schema with all the repeated strings (not sure if that is critical though, but it would get worse if the 'any' list had `optional` and `sequence` types added). If specifying individual constraints are in fact optional it's also lightweight for the kernel registration - only specify the constraints on inputs/outputs you need them for.

Reply:
Sounds good. If a Dev wants to implement a new Op and take the dynamic schema, Dev-self should handle the type-constraint in the kernel implementation and check functions.